creator-node/src/serviceRegistry.js

```diff
@@ -44,64 +52,174 @@ class ServiceRegistry {
+      this.logInfo(`Attempting to recover node L1 identity for endpoint ${endpoint} || attempt #${++attempt} ...`)
+
+      try {
+        const spID = await this.libs.ethContracts.ServiceProviderFactoryClient.getServiceProviderIdFromEndpoint(
+          endpoint
+        )
+        this.nodeConfig.set('spID', spID)
+
+        // Swallow any errors during recovery attempt
+      } catch (e) {
+        this.logError(`RecoverNodeL1Identity Error: ${e}`)
       }
-    },
-    discoveryProviderConfig: AudiusLibs.configDiscoveryProvider(discoveryProviderWhitelist),
-    // If an identity service config is present, set up libs with the connection, otherwise do nothing
-    identityServiceConfig: identityService ? AudiusLibs.configIdentityService(identityService) : undefined,
-    isDebug: config.get('creatorNodeIsDebug')
-  })
-
-  await audiusLibs.init()
-  return audiusLibs
+
+      await utils.timeout(5000, false)
+    }
+
+    this.logInfo(`Successfully recovered node L1 identity for endpoint ${endpoint} on attempt #${attempt}. spID =  ${this.nodeConfig.get('spID')}`)
+  }
+
+  /**
+   * Wait until spID config is set, then attempt to register on L2 URSM with infinite retries
+   */
+  async _registerNodeOnL2URSM () {
+    const retryTimeoutMs = 10000 // 10sec
+
+    // Wait until spID config is set
+    while (this.nodeConfig.get('spID') === 0) {
+      this.logInfo(`Cannot register node on L2 URSM until L1 identity (spID) is set. Retrying in ${retryTimeoutMs}ms`)
+      await utils.timeout(retryTimeoutMs, false)
+    }
+
+    // Attempt to register on URSM with infinite retries
+    let registered = false
+    let attempt = 0
+    while (registered === false) {
+      this.logInfo(`Attempting to register node on L2 URSM || attempt #${++attempt} ...`)
+
+      try {
+        await this.URSMRegistrationManager.run()
+        registered = true
+
+        // Swallow any errors during registration attempt
+      } catch (e) {
+        this.logError(`RegisterNodeOnL2URSM Error: ${e}`)
+
+        if (e.message === 'URSMRegistration cannot run until UserReplicaSetManager contract is deployed') {
+          this.logInfo(`Checking if URSM has since been deployed...`)
+          await this.libs.contracts.initUserReplicaSetManagerClient(false)
+        }
+      }
+      await utils.timeout(retryTimeoutMs, false)
+    }
+
+    this.logInfo('URSM Registration completed')
+  }
+
+  /**
+   * Wait until spID config is set, then initialize SnapbackSM
+   */
+  async _initSnapbackSM () {
+    const retryTimeoutMs = 10000 // ms
+
+    while (this.nodeConfig.get('spID') === 0) {
+      this.logInfo(`Cannot initialize SnapbackSM until L1 identity (spID) is set. Retrying in ${retryTimeoutMs}ms`)
+      await utils.timeout(retryTimeoutMs, false)
+    }
+
+    let complete = false
+    while (complete === false) {
+      try {
+        this.snapbackSM = new SnapbackSM(this.nodeConfig, this.libs)
+        await this.snapbackSM.init()
```

Comment:
do these two things actually relate to each other? also can't seem to find the snapbackSM.init() function - where is that found?

Reply:
good catch, the constructor should have been called outside loop, fixed